2019/pkg/input/input.go

```diff
@@ -0,0 +1,22 @@
+package input
+
+import (
+	"bufio"
+	"os"
+)
+
+func ReadFromFile(filepath string) ([]string, error) {
```

Comment:
Because you are reading the whole input anyways, you can do:

```
data, err := os.ReadFile(filepath)
if err != nil { return nil, err }
return strings.Split(string(data), "\n"), nil
```

Reply:
Since I only have one use of `ReadFromFile` at the moment and it wants the output as `[]int`, I changed this method so that it combines reading from file and converting `Atoi`.  I named the func `ReadFromFileToI` which I don't like.  You have any thoughts on this?

Reply:
That sounds good as well. Although, I would rename it to something like "ReadCode" or similar, to make the purpose clearer.

Reply:
why "ReadCode"?  This new method is doing the usual thing of reading in each line of a file as a string, but then converting the string to an int. Maybe `ReadIntsFromFile` is better.

Reply:
That would be fine as well.

The reason I prefer `ReadCode`, is because it doesn't lose the "problem domain" -- so the purpose is clearer. You can read about a similar problem when searching "anemic domain model".

Reply:
I'm reading about "anemic domain model" on [wiki](https://en.wikipedia.org/wiki/Anemic_domain_model). In reading that, it seems there are 3 parts being discussed... the "domain object", "business logic", and the "program itself". In this example of `ReadCode`, I guess the "domain object" is this `ReadCode` method ... or is the "domain object" the object that consumes the output of `ReadCode`?  From reading the wiki, I don't see anything about naming methods and "problem domain". This anemic-domain-model more seems to be saying you should couple the business logic with the object that uses the logic. How does this relate to naming this function `ReadCode` vs `ReadIntsFromFile`?

Reply:
A lot of DDD information is Java centric -- so "object" is an implementation detail. But, the main idea is "preserve domain concepts/terminology in code, because this makes it easier to understand how things relate to the bigger problem". In this case you could say `ReadCode` is a "domain function/process" and `Code` is a "domain object/struct".

Coupling business logic with the object that uses the logic is somewhat an artifact of using Java. It also quite often is convenient. The coupling between the business logic and the object is also necessary -- however, it doesn't necessarily mean it needs to be a method on the object.

I can give you a similar example of "anemic code":

```
package domain
const TwentyFive = 0.26
func Calculate(v float64) float64 { return TwentyFive * v }
```

Here the code has forgotten everything about the domain it was about and why it exists. It might be improved with comments, but the naming can be improved:

```
package tax
const IncomeRate = 0.26
func IncomePaid(v float64) float64 { return IncomeRate * v }
```

(Ignore that it's nonsense in terms of actual domain.)

Of course the `ReadCode` and `ReadIntsFromFile` in the current situation doesn't make a big difference. But let's imagine the "code" format is further elaborated, e.g. it contains some additional directives that aren't ints.

Reply:
I agree that `ReadCode` and `ReadIntsFromFile` in the current situation doesn't make a big difference, but I think the discussion is interesting even if its a little overkill for the situation at hand... in part since these Advent of Code problems don't seem to have a domain... or the domain is about something silly like elfs fixing a broken engine lol. Besides all that I still think there is something to learn here and I like the discussion.

Generally, I like this idea of "preserve domain in code...". I think it would be fun to practice because I don't normally prioritize this concept when writing code. Usually when I write code my goals are to reduce complexity, reduce code duplication, and increase readability, and increase ability to test the code. I also aim to write code in a way that can be easily re-written in the future, decoupled and modular.

So I guess I kind of feel like "preserving the domain" is opposed to some of my other goals...like decoupling and reducing code duplication.

For example, the reason I put the `ReadIntsFromFile` in its own `input` package was because I was trying to de-dup the code and generalize it so that I can use it for many of the problems in Advent of Code sine many of the problems need to parse a file as input in a similar way. So I was planning on adding all the parsing logic in the input package. But that effort to de-deup the code does lose the domain specific info. I think with real production cases, data is more complex and the processing of that data being coupled to the domain makes sense.

I understand that I lose domain information when I generalize this input parsing method. So one question is... when should things be generalize to a generic package versus keeping the logic coupled with the domain?

Reply:
> or the domain is about something silly like elfs fixing a broken engine lol.

In this case I would consider the main domain the "engine virtual machine".

> So I guess I kind of feel like "preserving the domain" is opposed to some of my other goals...like decoupling and reducing code duplication.

Having "decoupling" and "reducing code duplication" as main metric can be conflicting. I see those as tools to achieve other things. I try to consider more "code clarity", "code maintainability" and "code correctness". As an example, "decoupling" by itself may reduce all of those three.

> So one question is... when should things be generalize to a generic package versus keeping the logic coupled with the domain?

One approach is [Rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)). In other words, if you refactor before having three examples where it will be used, you won't have enough information to create an appropriate abstraction.

Similarly, you can still preserve both, e.g.:
```
func ParseCodeFile(path string) (Code, error) {
     ints, err := input.ReadIntsFromFile(path)
     ...
     return Code(ints), nil
}
```

Of course, you would want to add that "ReadIntsFromFile", if it's significant amount of code and needs to be reused from multiple places.

"A Philosophy of Software Design" by John Ousterhout contains good advice for it.

Reply:
In the case of Rule of Three I'm going to remove the `input` package since `ReadIntsFromFile` is only being used in 1 place. Ty :)